File: tx_sysvar_js/buffer.js
const { LocalBlock } = require('./block_ingestion');

const BatchSize = 10;

class Buffer {
    constructor(clock, blockSvc, batchSize = 10) {
        this.txs = [];
        this.slot = 0;
        this.clock = clock;
        this.blockSvc = blockSvc;
        this.batchSize = batchSize
        this.firstTxTime = performance.now()
    }

    Add(slot, tx) {
        // console.log("Adding tx to buffer", slot);

        if (slot < this.slot) {
            return;
        }

        if (slot > this.slot || this.txs.length >= this.batchSize) {
            const temp = this.txs;
            const currentSlot = this.slot;

            // Go spawns a goroutine: go b.buildBlock(b.slot, temp)
            setImmediate(() => {
                this.buildBlock(currentSlot, temp);
            });

            this.txs = [];
            if (slot > this.slot) {
                const duration = (performance.now() - this.firstTxTime).toFixed(3);
                process.stdout.write(`Time from first tx -> block formed: ${duration}ms\n`);
                this.firstTxTime = performance.now()
                this.slot = slot;
            }
        }
        this.txs.push(tx);
    }

    buildBlock(slot, txs) {
        console.log(`Building Batch: ${slot} | Len: ${txs.length}`);
        const block = new LocalBlock(
            slot,
            this.clock.TimeStamp(slot),
            Date.now(),
            txs
        );
        this.blockSvc.Queue(block);
    }
}

module.exports = { Buffer };

File: tx_sysvar/buffer.go
package main

import (
	"fmt"
	"log"
	"sync"
	"time"

	pb "github.com/andrew-solarstorm/yellowstone-grpc-client-go/proto"
)

const BatchSize int8 = 10

type Buffer struct {
	mu          *sync.RWMutex
	txs         []*pb.SubscribeUpdateTransactionInfo
	firstTxTime time.Time

	slot uint64

	clock    *SystemClock
	blockSvc *BlockBuilder
}

func (b *Buffer) Add(slot uint64, tx *pb.SubscribeUpdateTransactionInfo) {
	// log.Println("Adding tx to buffer", slot)
	b.mu.Lock()
	defer b.mu.Unlock()
	if slot < b.slot {
		return
	}

	if slot > b.slot || len(b.txs) >= int(BatchSize) {
		temp := b.txs
		go b.buildBlock(b.slot, temp)

		b.txs = make([]*pb.SubscribeUpdateTransactionInfo, 0)
		if slot > b.slot {
			b.slot = slot
			fmt.Printf("Time from first tx -> block formed: %s\n", time.Since(b.firstTxTime).String())
			b.firstTxTime = time.Now()
		}

	}
	if tx != nil {
		b.txs = append(b.txs, tx)
	}
}

func (b *Buffer) buildBlock(slot uint64, txs []*pb.SubscribeUpdateTransactionInfo) {
	log.Printf("Building Batch: %d | Len: %d\n", slot, len(txs))
	block := &LocalBlock{
		Slot:         slot,
		Blocktime:    b.clock.TimeStamp(slot),
		CreatedAt:    time.Now().UnixMilli(),
		Transactions: txs,
	}
	b.blockSvc.Queue(block)
}
